xrpl4j-crypto-parent/xrpl4j-crypto-bouncycastle/src/main/java/org/xrpl/xrpl4j/crypto/signing/DerivedKeysSignatureService.java

```diff
@@ -0,0 +1,210 @@
+import org.xrpl.xrpl4j.crypto.KeyMetadata;
+import org.xrpl.xrpl4j.crypto.KeyStoreType;
+import org.xrpl.xrpl4j.crypto.PrivateKey;
+import org.xrpl.xrpl4j.crypto.PublicKey;
+import org.xrpl.xrpl4j.crypto.Seed;
+import org.xrpl.xrpl4j.crypto.ServerSecretSupplier;
+import org.xrpl.xrpl4j.keypairs.DefaultKeyPairService;
+import org.xrpl.xrpl4j.keypairs.KeyPair;
+import org.xrpl.xrpl4j.keypairs.KeyPairService;
+import org.xrpl.xrpl4j.model.transactions.Transaction;
+
+import java.util.Arrays;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * An implementation that uses an in-memory secret key in order to deterministically create a seed value that can then
+ * be used to deterministically generate XRPL private keys. Any derived key can then be used for signing XRP
+ * transactions. This implementation keeps a cache of instances of {@link SingleKeySignatureService} that it delegates
+ * to based upon {@link KeyMetadata} supplied on each call.
+ *
+ * <p>WARNING: This implementation stores private seed-generation material in-memory, and is thus only meant for
+ * lower-security environments. For higher security deployments, prefer an HSM-based implementation instead.</p>
+ */
+public class DerivedKeysSignatureService implements SignatureService {
+
+  private static final byte[] EMPTY_BYTE_ARRAY = new byte[0];
+
+  private final VersionType versionType;
+  private final KeyPairService keyPairService;
+
+  private final LoadingCache<KeyMetadata, SingleKeySignatureService> keyMetadataLoadingCache;
+
+  private final ServerSecretSupplier serverSecretSupplier;
+
+  /**
+   * Required-args Constructor for use in development mode.
+   *
+   * @param serverSecretSupplier A {@link ServerSecretSupplier} that can be used to generate seed values, which can then
+   *                             be used to generate private keys.
+   * @param versionType          A {@link VersionType} that defines which type of key this signature service uses.
+   */
+  public DerivedKeysSignatureService(
+    final ServerSecretSupplier serverSecretSupplier,
+    final VersionType versionType
+  ) {
+    this(serverSecretSupplier, versionType, new DefaultKeyPairService());
+  }
+
+  /**
+   * Required-args Constructor.
+   *
+   * @param serverSecretSupplier A {@link ServerSecretSupplier} that can be used to generate seed values, which can
+   * @param versionType          A {@link VersionType} that defines which type of key this signature service uses.
+   * @param keyPairService       A {@link KeyPairService}.
+   */
+  public DerivedKeysSignatureService(
+    final ServerSecretSupplier serverSecretSupplier,
+    final VersionType versionType,
+    final KeyPairService keyPairService
+  ) {
+    this.serverSecretSupplier = Objects.requireNonNull(serverSecretSupplier);
+    this.versionType = Objects.requireNonNull(versionType);
+    this.keyPairService = Objects.requireNonNull(keyPairService);
+
+    this.keyMetadataLoadingCache = Caffeine.newBuilder()
```

Comment:
Can we make this configurable?

Reply:
Captured in https://github.com/XRPLF/xrpl4j/issues/58.